client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java

```diff
@@ -33,15 +35,31 @@
 
     private static final PositionSerializer SERIALIZER = new PositionSerializer();
     private final Map<Segment, Long> ownedSegments;
+    private final Map<Segment, Range> segmentRanges;
 
     /**
      * Instantiates Position with current and future owned segments.
      *
-     * @param ownedSegments Current segments that the position refers to.
+     * @param segments Current segments that the position refers to.
      */
+    public PositionImpl(Map<SegmentWithRange, Long> segments) {
+        this.ownedSegments = new HashMap<>(segments.size());
+        this.segmentRanges = new HashMap<>(segments.size());
+        for (Entry<SegmentWithRange, Long> entry : segments.entrySet()) {
+            SegmentWithRange s = entry.getKey();
+            this.ownedSegments.put(s.getSegment(), entry.getValue());
+            this.segmentRanges.put(s.getSegment(), s.getRange());
+        }
+    }
+    
     @Builder(builderClassName = "PositionBuilder")
-    public PositionImpl(Map<Segment, Long> ownedSegments) {
-        this.ownedSegments = new HashMap<>(ownedSegments);
+    private PositionImpl(Map<Segment, Long> ownedSegments, Map<Segment, Range> segmentRanges) {
```

Comment:
Should we continue to keep this constructor? 
The position object created from different constructors will either have the range information or not. Which can make it ambiguous for the user of this object. 

Why not keep the range information by default and only have that constructor?

Reply:
There already is only one _public_ constructor which requires range information. This call is private. It is needed for the codegened builder which is in-turn used in the PositionSerializer below (Which is also private). It has to be a builder due to the interface defined by the serialization framework.